barch.1
.\" $Id$
.TH barch 1
.SH NAME
barch \- backup archive utility
.SH SYNOPSIS
.B barch
.I <operation>
.I [options]
.SH DESCRIPTION
barch is an archive utility specifically designed with the needs of the
backup and restore process in mind.
.SH OPTIONS
Most of the following options were adopted from
.B tar
in order to ease migration from one tool to the other.
.SS Operation Modes
.TP
.B -c, --create
Create a new archive.
.TP
.B -t, --list
List the contents of an archive.
.TP
.B -x, --extract
Extract files from an archive.
.SS Operation Modifiers
.TP
.B -G, --incremental
When creating an archive, record all of the files encountered in each
matching directory in the archive file, for use when extracting the
archive.  When extracting files from an archive, the list of files in
each matching directory is compared against the files in the
corresponding destination directory.  All files present in the
destination that are not listed in the archive are removed.
.TP
.B -g FILE, --listed-incremental=FILE
Enables incremental mode as described above, but also records a snapshot
of the time when the archive was started and all directories traversed
into the named file.  If the snapshot file previously existed, the
contents of the output archive are limited to files that have been
modified in any way (including being renamed) more recently than the
timestamp in the file and the entire contents of directories that have
been renamed or replaced.
.TP
.B --differential=FILE
Similar to the listed-incremental mode above, this mode uses the
snapshot file to determine what files to archive, but does not update
the file when the backup process is complete.
.TP
.B --overwrite-files
When extracting an archive, overwrite files directly instead of using
temporary files.
.TP
.B -j, --bzip2
Filter (compress or uncompress) the archive using
.BR bzip2 .
.TP
.B -z, --gzip
Filter (compress or uncompress) the archive using
.BR gzip .
.TP
.B -Z, --compress
Filter (compress or uncompress) the archive using
.BR compress .
.SS Archive I/O Options
.TP
.B -f FILE, --file=FILE
Archive filename (defaults to standard input when reading archives, or
standard output when writing archives).
.SS Local File Selection
.TP
.B -C PATH, --directory=PATH
Change to named directory after opening archive and before reading or
writing files.
.TP
.B --exclude=PATTERN
File names to exclude files, given as a glob pattern.
.TP
.B -X FILE, --exclude-from=FILE
Exclude patterns listed in the named file.
.TP
.B --include=PATTERN
Include files, given as a glob pattern
.TP
.B -I FILE, --include-from=FILE
Include patterns listed in the named file
.TP
.B --patterns-from=FILE
Ex/include patterns listed in the named file.  Each line in the file
contains one pattern, prefixed by either
.B +
for files to include, or
.B -
for files to exclude.  Lines starting with
.B #
are ignored.  The patterns are applied in order, and the first matching
pattern determines the outcome.
.TP
.B -P, --absolute-names
Don't strip leading '/'s from file names
.TP
.B -h, --dereference
Dump instead the files symlinks point to
.TP
.B -N FILE, --newer=FILE, --after-date=FILE
Only store files newer than the given file
.TP
.B --one-file-system
When creating an archive, do not recurse into directories that are
located on different filesystems from their parents.  The directory
itself is still recorded in the archive, for compatibility with
.BR tar .
Note that directories specifically listed on the command line are always
archived.
.SS Informative Output
.TP
.B -v, --verbose
Show more information about progress.  This option may be used multiple
times.
.TP
.B --checkpoint
Print directory names as they are processed.
.TP
.B --totals
Output the total number of bytes written after the archive is created.
.SH WARNINGS
The default behavior of
.B barch
is to create temporary files and then rename them on top of the
destination file.  This default behavior is required to overwrite files
that are in use (ie mmaped), but it destroys any hard links to files
being written.  Use
.B --overwrite-files
if this is a problem.


NEWS
-------------------------------------------------------------------------------
Changes in version 0.6

- Renamed --snapshot to --listed-incremental and added -G option to make
  the incremental mode specification identical to that of tar.

- Added --differential mode, which works identically to
  --listed-incremental but without updating the listing file.

- Fixed a bug in incremental backups where excluded files would still be
  listed (but not recorded) in the archive.

- Fixed a couple of minor bugs in extracting incremental backups.

- Added --remove-files option.

- Added full support for extracting device nodes.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.5

Since the last release, barch has been completely rewritten in C
(instead of C++).  The resulting source is cleaner and easier to develop
(not to mention at least slightly more efficient).  The resulting
program is also more feature complete, and less of a prototype work.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------


barch.c
/* $Id$ */

#include <sysdeps.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <sys/mman.h>

#include <adt/ghash.h>
#include <cli/cli.h>
#include <crc/crc64.h>
#include <iobuf/iobuf.h>
#include <msg/msg.h>
#include <str/iter.h>
#include <str/str.h>

#include "barch.h"

static void add_exinclude(const char*, const struct cli_option*);

const char stamp[8] = "=BArch3\n";
//const char stamp[8] = { 0xba,0xee,0x81,0xb7,0xe6,0x83,0xbf, FILE_VERSION };

int opt_absolute = 0;
int opt_checkpoint = 0;
int opt_dereference = 0;
const char* opt_filename = "-";
int opt_incremental = 0;
int opt_iobuflen = 64*1024;
const char* opt_snapshot = 0;
const char* opt_differential = 0;
int opt_usetmp = 1;
int opt_verbose = 0;
time_t opt_timestamp = 0;
int opt_totals = 0;
int opt_compress = 0;
int opt_onefilesystem = 0;
int opt_removefiles = 0;

static int cmd_create = 0;
static int cmd_list = 0;
static int cmd_extract = 0;
static const char* opt_directory = 0;
static const char* opt_timestr = 0;

const char program[] = "barch";
const int msg_show_pid = 0;
const char cli_help_prefix[] = "Create a barch archive\n";
const char cli_help_suffix[] = "";
const char cli_args_usage[] = "[file ...]";
const int cli_args_min = 0;
const int cli_args_max = -1;
cli_option cli_options[] = {
  { 0, "Main operation mode", CLI_SEPARATOR, 0, 0, 0, 0 },
  { 'c', "create", CLI_FLAG, 1, &cmd_create,
    "Create a new archive", 0 },
  { 't', "list", CLI_FLAG, 1, &cmd_list,
    "List the contents of an archive", 0 },
  { 'x', "extract", CLI_FLAG, 1, &cmd_extract,
    "Extract files from an archive", 0 },

  { 0, "Operation modifiers", CLI_SEPARATOR, 0, 0, 0, 0 },
  { 'G', "incremental", CLI_FLAG, 1, &opt_incremental,
    "Incremental backup or restore", 0 },
  { 'g', "listed-incremental", CLI_STRING, 0, &opt_snapshot,
    "Listed incremental backup (updates the file)", 0 },
  { 0,   "differential", CLI_STRING, 0, &opt_differential,
    "Listed incremental backup (no update)", 0 },
  { 0, "overwrite-files", CLI_FLAG, 0, &opt_usetmp,
    "Don't extract to temporary files first", 0 },
  { 0, "remove-files", CLI_FLAG, 1, &opt_removefiles,
    "Remove files after adding them to archive", 0 },
  { 'j', "bzip2", CLI_FLAG, COMPRESS_BZ2, &opt_compress,
    "Filter the archive through bzip2", 0 },
  { 'z', "gzip", CLI_FLAG, COMPRESS_GZ, &opt_compress,
    "Filter the archive through gzip", 0 },
  { 'Z', "compress", CLI_FLAG, COMPRESS_Z, &opt_compress,
    "Filter the archive through compress", 0 },

#if 0
  // FIXME: handle all these options
  { 0,   "Handling of file attributes", CLI_SEPARATOR, 0, 0, 0, 0 },
  { 0,   "owner", CLI_STRING, 0, &opt_owner,
    "Force owner for added files", 0 },
  { 0,   "group", CLI_STRING, 0, &opt_group,
    "Force group for added files", 0 },
  { 'm', "modification-time", CLI_FLAG, 1, &opt_nomtime,
    "Don't extract file modified time", 0 },
  { 0,   "same-owner", CLI_FLAG, 1, &opt_chown,
    "Try extracting files with the same ownership", 0 },
  { 0,   "no-same-owner", CLI_FLAG, 0, &opt_chown,
    "Extract files as yourself", 0 },
  { 'p', "same-permissions", CLI_FLAG, 1, &opt_chmod,
    "Extract permissions information", 0 },
  { 0,   "no-same-permissions", CLI_FLAG, 0, &opt_chmod,
    "Do not extract permissions information", 0 },
  { 0,   "preserve-permissions", CLI_FLAG, 1, &opt_chmod,
    "Same as -p", 0 },
  //
#endif

  { 0, "Archive input/output options", CLI_SEPARATOR, 0, 0, 0, 0 },
  { 'f', "file", CLI_STRING, 0, &opt_filename,
    "Archive filename", "standard input/output" },

  { 0, "Local file selection", CLI_SEPARATOR, 0, 0, 0, 0 },
  { 'C', "directory", CLI_STRING, 0, &opt_directory,
    "Change to named directory after opening archive", 0 },
  { 0, "exclude", CLI_FUNCTION, 0, add_exinclude,
    "Exclude files, given as a glob pattern", 0 },
  { 'X', "exclude-from", CLI_FUNCTION, 0, add_exinclude,
    "Exclude patterns listed in the named file", 0 },
  { 0, "include", CLI_FUNCTION, 0, add_exinclude,
    "Include files, given as a glob pattern", 0 },
  { 'I', "include-from", CLI_FUNCTION, 0, add_exinclude,
    "Include patterns listed in the named file", 0 },
  { 0, "patterns-from", CLI_FUNCTION, 0, add_exinclude,
    "Ex/include patterns listed in the named file", 0 },
  { 'P', "absolute-names", CLI_FLAG, 1, &opt_absolute,
    "Don't strip leading '/'s from file names", 0 },
  { 'h', "dereference", CLI_FLAG, 1, &opt_dereference,
    "Dump instead the files symlinks point to", 0 },
  { 'N', "newer", CLI_STRING, 0, &opt_timestr,
    "Only store files newer than the given file", 0 },
  { 0, "after-date", CLI_STRING, 0, &opt_timestr,
    "Same as --newer", 0 },
  { 0, "one-file-system", CLI_FLAG, 1, &opt_onefilesystem,
    "Do not recurse into different file systems", 0 },

  { 0, "Informative output", CLI_SEPARATOR, 0, 0, 0, 0 },
  { 'v', "verbose", CLI_COUNTER, 1, &opt_verbose,
    "Show more information about progress", 0 },
  { 0, "checkpoint", CLI_FLAG, 1, &opt_checkpoint,
    "Print directory names as they are processed", 0 },
  { 0, "totals", CLI_FLAG, 1, &opt_totals,
    "Print total bytes written when creating archive", 0 },
  {0,0,0,0,0,0,0}
};

pid_t pid;
struct timeval start_time;

static int cwd_fd;

void do_chdir(void)
{
  if (opt_directory) {
    if ((cwd_fd = open(".", O_RDONLY)) == -1)
      die1sys(1, "Could not open current directory");
    if (chdir(opt_directory) != 0)
      die3sys(1, "Could not change directory to '", opt_directory, "'");
  }
}

void chdir_back(void)
{
  if (cwd_fd != 0 && fchdir(cwd_fd) != 0)
    die1sys(1, "Could not change directory back");
}

static void parse_timestamp(void)
{
  struct stat st;
  if (opt_timestr == 0) return;
  if (stat(opt_timestr, &st) != 0)
    die3sys(1, "Could not stat '", opt_timestr, "'");
  opt_timestamp = st.st_mtime;
}

char* iobuffer;
unsigned long iobuflen;
static int iobufmmap;

static void alloc_iobuf(void)
{
  if ((iobuffer = mmap(0, iobuflen, PROT_READ|PROT_WRITE,
		    MAP_PRIVATE|MAP_ANON, -1, 0)) != MAP_FAILED)
    iobufmmap = 1;
  else {
    if ((iobuffer = malloc(iobuflen)) != 0)
      iobufmmap = 0;
    else
      die_oom(1);
  }
}

static void make_iobuf(void)
{
  iobuflen = getpagesize();
  while ((long)iobuflen < opt_iobuflen) iobuflen *= 2;
  alloc_iobuf();
}

void ready_iobuf(unsigned long size)
{
  if (iobuflen >= size) return;
  if (iobufmmap) munmap(iobuffer, iobuflen);
  else free(iobuffer);
  while (iobuflen < size) iobuflen *= 2;
  alloc_iobuf();
}

/* Excludes/Includes *********************************************************/
static str exincludes = {0,0,0};

static void add_exinclude_one(char t, const char* s)
{
  if (exincludes.len > 0)
    if (!str_catc(&exincludes, 0)) die_oom(1);
  if (t != 0)
    if (!str_catc(&exincludes, t)) die_oom(1);
  if (!str_cats(&exincludes, s)) die_oom(1);
}

static void add_exinclude_file(char t, const char* f)
{
  ibuf in;
  str line = {0,0,0};
  if (!ibuf_open(&in, f, 0))
    die3sys(1, "Could not open '", f, "'");
  while (ibuf_getstr(&in, &line, LF)) {
    str_strip(&line);
    if (line.len > 0 && line.s[0] != '#')
      add_exinclude_one(t, line.s);
  }
  str_free(&line);
  ibuf_close(&in);
}

static void add_exinclude(const char* s, const struct cli_option* o)
{
  if (o->name[0] == 'p')
    add_exinclude_file(0, s);
  else {
    char t = (o->name[0] == 'e') ? '-' : '+';
    if (o->ch != 0)
      add_exinclude_file(t, s);
    else
      add_exinclude_one(t, s);
  }
}

static int check_filename_part(const char* path)
{
  striter i;
  const str p = { (char*)path, strlen(path), 0 };
  for (striter_start(&i, &exincludes, 0);
       striter_valid(&i);
       striter_advance(&i))
    if (str_globs(&p, i.startptr+1))
      return *i.startptr == '+';
  return -1;
}

static int is_prefix(const char* path, const char* arg)
{
  unsigned long arglen = strlen(arg);
  unsigned long pathlen = strlen(path);
  while (arglen > 1 && arg[arglen-1] == '/') --arglen;
  if (arglen > pathlen) return 0;
  if (memcmp(arg, path, arglen) != 0) return 0;
  return arglen == pathlen || path[arglen] == '/';
}

int check_filename(const char* path, int argc, char* argv[])
{
  int r;
  if (argc > 0) {
    for (r = 0; r < argc; ++r)
      if (is_prefix(path, argv[r]))
	return 1;
    return 0;
  }
  if ((r = check_filename_part(path)) >= 0) return r;
  if ((path = strrchr(path, '/')) != 0)
    if ((r = check_filename_part(++path)) >= 0) return r;
  return 1;
}

/* Main routine **************************************************************/
extern int do_create(int, char**);
extern int do_list(int, char**);
extern int do_extract(int, char**);

int cli_main(int argc, char* argv[])
{
  if (cmd_create + cmd_list + cmd_extract != 1)
    die1(1, "Must specify exactly one of create, list, or extract");

  pid = getpid();
  gettimeofday(&start_time, 0);
  pwcache_init();
  parse_timestamp();
  make_iobuf();
  if (opt_snapshot != 0)
    opt_differential = opt_snapshot;
  if (opt_differential != 0)
    opt_incremental = 1;

  if (cmd_create) return do_create(argc, argv);
  if (cmd_list) return do_list(argc, argv);
  if (cmd_extract) return do_extract(argc, argv);
  return 1;
}
